File: app/src/main/java/uk/cybertecpro/cyberlearningos/LearningStore.kt
package uk.cybertecpro.cyberlearningos

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject

class LearningStore(context: Context) {
    private val preferences = context.getSharedPreferences("learning-os-v1", Context.MODE_PRIVATE)

    fun load(): List<LearningTopic> {
        val raw = preferences.getString("topics", "[]") ?: "[]"
        return runCatching {
            val array = JSONArray(raw)
            List(array.length()) { index -> fromJson(array.getJSONObject(index)) }
        }.getOrDefault(emptyList())
    }

    fun save(topics: List<LearningTopic>) {
        val array = JSONArray()
        topics.forEach { array.put(toJson(it)) }
        preferences.edit().putString("topics", array.toString()).apply()
    }

    private fun toJson(topic: LearningTopic) = JSONObject().apply {
        put("id", topic.id)
        put("title", topic.title)
        put("purpose", topic.purpose)
        put("capability", topic.capability)
        put("stage", topic.stage.name)
        put("primeGist", topic.primeGist)
        put("coreNotes", topic.coreNotes)
        put("connections", topic.connections)
        put("retrieval", topic.retrieval)
        put("application", topic.application)
        put("analystExplanation", topic.analystExplanation)
        put("leaderExplanation", topic.leaderExplanation)
        put("executiveExplanation", topic.executiveExplanation)
        put("feedback", topic.feedback)
        put("createdAt", topic.createdAt)
        put("dueAt", topic.dueAt)
        put("intervalDays", topic.intervalDays)
    }

    private fun fromJson(json: JSONObject) = LearningTopic(
        id = json.optString("id"),
        title = json.optString("title"),
        purpose = json.optString("purpose"),
        capability = json.optString("capability"),
        stage = runCatching { LearningStage.valueOf(json.optString("stage")) }
            .getOrDefault(LearningStage.PRIME),
        primeGist = json.optString("primeGist"),
        coreNotes = json.optString("coreNotes"),
        connections = json.optString("connections"),
        retrieval = json.optString("retrieval"),
        application = json.optString("application"),
        analystExplanation = json.optString("analystExplanation"),
        leaderExplanation = json.optString("leaderExplanation"),
        executiveExplanation = json.optString("executiveExplanation"),
        feedback = json.optString("feedback"),
        createdAt = json.optLong("createdAt", System.currentTimeMillis()),
        dueAt = json.optLong("dueAt"),
        intervalDays = json.optInt("intervalDays"),
    )
}

File: windows/CyberLearningOS.Windows/LearningStore.cs
using System.IO;
using System.Text.Json;

namespace CyberLearningOS.Windows;

public sealed class LearningStore
{
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        WriteIndented = true,
    };

    private readonly string _path;

    public LearningStore()
    {
        var folder = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), "CyberLearningOS");
        Directory.CreateDirectory(folder);
        _path = Path.Combine(folder, "topics.json");
    }

    public IReadOnlyList<LearningTopic> Load()
    {
        try
        {
            if (!File.Exists(_path)) return [];
            var raw = File.ReadAllText(_path);
            var topics = JsonSerializer.Deserialize<List<LearningTopic>>(raw, JsonOptions) ?? [];
            using var document = JsonDocument.Parse(raw);
            for (var index = 0; index < topics.Count; index++)
            {
                var element = document.RootElement[index];
                if (!element.TryGetProperty("currentStep", out _)) MigrateLegacy(topics[index], element);
                Normalise(topics[index]);
            }
            return topics;
        }
        catch (JsonException)
        {
            return [];
        }
    }

    public void Save(IEnumerable<LearningTopic> topics)
    {
        var temp = _path + ".tmp";
        File.WriteAllText(temp, JsonSerializer.Serialize(topics, JsonOptions));
        File.Move(temp, _path, true);
    }

    private static void Normalise(LearningTopic topic)
    {
        topic.CurrentStep = Math.Clamp(topic.CurrentStep, 0, LearningGuide.Steps.Count - 1);
        topic.StepEvidence ??= [];
        while (topic.StepEvidence.Count < LearningGuide.Steps.Count) topic.StepEvidence.Add("");
        if (topic.StepEvidence.Count > LearningGuide.Steps.Count)
            topic.StepEvidence.RemoveRange(LearningGuide.Steps.Count, topic.StepEvidence.Count - LearningGuide.Steps.Count);
    }

    private static void MigrateLegacy(LearningTopic topic, JsonElement element)
    {
        var stage = element.TryGetProperty("stage", out var stageElement) ? stageElement.GetInt32() : 0;
        topic.CurrentStep = stage switch
        {
            0 => 3,
            1 => 4,
            2 => 6,
            3 => 12,
            4 => 10,
            5 => 13,
            6 => 11,
            7 => 13,
            _ => 0,
        };
        topic.Completed = stage == 7;
        topic.StepEvidence = Enumerable.Repeat("", LearningGuide.Steps.Count).ToList();
        topic.StepEvidence[0] = topic.Capability;
        topic.StepEvidence[1] = topic.Purpose;
        topic.StepEvidence[3] = Value(element, "primeGist");
        topic.StepEvidence[6] = Value(element, "connections");
        topic.StepEvidence[7] = Value(element, "coreNotes");
        topic.StepEvidence[10] = Value(element, "application");
        topic.StepEvidence[11] = Value(element, "feedback");
        topic.StepEvidence[12] = Value(element, "retrieval");
        topic.StepEvidence[13] = string.Join("\n\n", new[]
        {
            Value(element, "analystExplanation"),
            Value(element, "leaderExplanation"),
            Value(element, "executiveExplanation"),
        }.Where(value => !string.IsNullOrWhiteSpace(value)));
    }

    private static string Value(JsonElement element, string name) =>
        element.TryGetProperty(name, out var value) ? value.GetString() ?? "" : "";
}
